### PR TITLE
Tandem Curriculum Ramp: smooth linear ramp instead of hard cutoff

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current data curriculum skips tandem samples for the first 10 epochs, then reintroduces them at FULL weight. This abrupt switch creates a distribution shift shock — the gradient signal suddenly changes from single-foil-only to a mixed distribution. PR #768 student notes specifically observed instability at this transition point.

A **smooth linear ramp** from weight 0 to 1.0 over epochs 10-30 allows the model to adapt incrementally to tandem samples, maintaining stable optimization. This is well-supported by curriculum learning theory (Bengio et al., ICML 2009).

This is a pure training dynamics change — no new parameters, no architecture modification, no new features. The endpoint (full tandem weight at epoch 30) is the same as current behavior; only the path changes.

## Instructions

Add `--tandem_curriculum_ramp` flag to replace the abrupt reintroduction with a smooth ramp.

### Step 1: Add argument

```python
parser.add_argument('--tandem_curriculum_ramp', action='store_true',
                    help='Smooth linear ramp for tandem sample weights (epochs 10-30) instead of hard cutoff')
```

### Step 2: Find the tandem skip logic

Find where tandem samples are skipped during early training. Look for code like:
```python
if epoch < 10:
    # skip tandem samples or set their weight to 0
```

### Step 3: Replace with smooth ramp

```python
if args.tandem_curriculum_ramp:
    if epoch < 10:
        tandem_ramp_weight = 0.0
    elif epoch < 30:
        tandem_ramp_weight = (epoch - 10) / 20.0  # Linear: 0→1 over epochs 10-30
    else:
        tandem_ramp_weight = 1.0
else:
    # Current behavior: hard cutoff at epoch 10
    tandem_ramp_weight = 0.0 if epoch < 10 else 1.0
```

### Step 4: Apply the ramp weight

Multiply ALL tandem sample loss contributions by `tandem_ramp_weight`:

```python
# In the loss computation, where tandem samples are identified:
if is_tandem:
    loss = loss * tandem_ramp_weight
```

**Important:** The existing 1.5× tandem surface boost should still apply AFTER the ramp. So at epoch 30+, tandem samples get the full 1.5× boost. During the ramp period, they get `tandem_ramp_weight × 1.5×`.

### Step 5: Log the ramp weight

```python
if args.tandem_curriculum_ramp:
    wandb.log({"train/tandem_ramp_weight": tandem_ramp_weight})
```

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --seed 42 \
  --wandb_name "tanjiro/tandem-ramp-s42" \
  --wandb_group "tandem-curriculum-ramp" \
  --tandem_curriculum_ramp \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same flags, --seed 73, --wandb_name "tanjiro/tandem-ramp-s73"
```

## Baseline

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Your tandem curriculum ramp runs are catastrophically bad — p_in=50, p_tan=51 (vs baseline 11.7, 27.9). best_val_loss=0.89 suggests the model barely trained. The runs also finished in ~34 minutes instead of ~180 minutes.

Root cause is likely one of:
1. **The ramp weight was applied to ALL loss contributions, not just tandem.** Ensure the ramp weight ONLY multiplies tandem sample losses. Non-tandem samples must remain at full weight throughout training.
2. **The ramp weight zeroed out too much of the training data.** Check that single-foil samples are never affected by the ramp.
3. **Early training divergence.** If the curriculum somehow prevents gradient flow during epochs 0-10, the model can get stuck in a bad local minimum.

Debug steps:
- Run a quick 5-epoch debug with `--tandem_curriculum_ramp` and print the ramp weight and number of tandem vs non-tandem samples at each epoch.
- Verify that at epoch 0, tandem samples have weight=0 but ALL single-foil samples have weight=1.0.
- Verify that at epoch 30+, tandem samples also have weight=1.0.
- Check if the existing `tandem_loss_weight=1.5` multiplier is being combined correctly with the ramp weight (should be `1.5 * ramp_weight` for tandem samples).

The fix should be surgical — only the tandem loss path should be affected by the ramp.

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

**Re: advisor comment about catastrophic initial runs** — The initial runs (rfhq9u5n, udgy2wl8) hit the 30-minute wall-clock limit because I accidentally made train.py read the `SENPAI_TIMEOUT_MINUTES=30` env var instead of using the hardcoded 180-min timeout. This has been reverted. The re-runs below use the correct 180-min timeout and trained ~148 epochs like the baseline.

### Tandem Curriculum Ramp (epochs 10→30 linear ramp)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 11.34 | 11.65 | **11.50** | 11.742 | **-2.1%** ✅ |
| p_oodc | 8.1 | 7.6 | **7.85** | 7.643 | +2.7% ❌ |
| **p_tan** | 28.6 | 27.9 | **28.25** | 27.874 | +1.3% ❌ |
| p_re | 6.5 | 6.5 | **6.50** | 6.419 | +1.3% ❌ |

### Surface MAE (Ux, Uy) — val_in_dist best checkpoint

| Metric | Seed 42 | Seed 73 |
|--------|---------|---------|
| surf_Ux | 2.70 | 2.62 |
| surf_Uy | 0.89 | 0.81 |

### Volume MAE — val_in_dist best checkpoint

| Metric | Seed 42 | Seed 73 |
|--------|---------|---------|
| vol_Ux | 0.485 | 0.507 |
| vol_Uy | 0.192 | 0.188 |
| vol_p | 11.70 | 12.24 |

### Experiment details

- **Commands:** Exact commands from PR instructions (seed 42 and seed 73)
- **Epochs:** 147 (s42), 148 (s73) — 180-min wall-clock
- **Peak memory:** 46.1 GB (s42), 46.4 GB (s73)
- **W&B runs:** 993g849g (seed 42), p6gsbyy2 (seed 73)
- **W&B group:** tandem-curriculum-ramp

### What happened

**Mixed result — p_in improves but p_tan, p_oodc, p_re all regress.**

The smooth tandem ramp (0→1 over epochs 10-30) gives a meaningful -2.1% improvement on p_in (in-distribution pressure), which is the strongest improvement we've seen on p_in recently. However, this comes at the cost of regressions on all three other metrics: p_tan +1.3%, p_oodc +2.7%, p_re +1.3%.

**Interpretation:** The ramp successfully smooths the tandem reintroduction, which helps the model generalize better on in-distribution data. But the delayed exposure to full tandem weights (ramp doesn't hit 1.0 until epoch 30, which is ~20% of training) appears to reduce the model's ability to generalize to out-of-distribution conditions and Reynolds numbers. The p_oodc regression (+2.7%) is the largest and suggests the model doesn't get enough diverse tandem training signal early enough.

The seed variance is notable: seed 42 has p_tan=28.6 while seed 73 has p_tan=27.9, suggesting the tandem metric is sensitive to the ramp timing. Similarly, p_oodc varies from 8.1 (s42) to 7.6 (s73).

### Suggested follow-ups

1. **Shorter ramp (epochs 10-20):** The current 20-epoch ramp may be too slow. A 10-epoch ramp (10→20) would give the model earlier access to full tandem weights while still smoothing the transition.
2. **Warm ramp start:** Instead of starting from 0 at epoch 10, start from 0.3 or 0.5 — never fully zero out tandem samples, just downweight them during early training.
3. **Ramp only the tandem surface boost (1.5×), not the tandem inclusion:** Keep tandem samples at full loss weight throughout but ramp the 1.5× surface boost factor. This targets the distribution shift from the boost rather than the samples themselves.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2315 because ADVISOR: Closing — the tandem curriculum ramp shows a clear p_in improvement (-2.1%) but regresses the other 3 metrics: p_oodc +2.7%, p_tan +1.3%, p_re +1.3%. Net effect is negative.

The fundamental tradeoff is inherent: delaying tandem exposure helps in-distribution generalization at the cost of OOD/tandem/Re performance. Since p_tan is our gatekeeper metric and p_oodc has the largest gap to ensemble, we can't accept regressions there for p_in gains alone.

Your analysis was excellent — the observation that the ramp prevents early diverse tandem signal from shaping the representation is on point. The seed sensitivity (p_tan varies 27.9-28.6) also confirms this is a fragile approach.

Thank you for the thorough debugging of the initial catastrophic runs and the clean re-run. Moving you to a fresh experiment..
